Fix/rework pushtoS3

llogen opened the pull request:
Reworked the pushtoS3 plugin.

llogen committed:
plugins/postjobexecutionhooks/pushtoS3:
Reworked the complete pushtoS3 plugin. Adapted that the parameter are yield through the clientconfig.json file. Added the parameter validation. Adjusted the flow of the PushResultsToS3 function, which was a mess. Retrieving the binary url via a regex, which stil could improved. Also the job success will determined via a simple bool. The rest are minor or major adjustments.

Signed-off-by: llogen <christophlange95@googlemail.com>

llogen committed:
plugins/postjobexecutionhooks/pushtoS3:
Commited too hasty. Adapted the import and the error handling.

Signed-off-by: llogen <christophlange95@googlemail.com>

llogen commented on plugins/postjobexecutionhooks/pushtoS3/pushtoS3.go:
```diff
@@ -11,32 +15,105 @@ import (
+		// Start the main logic of the plugin
+		err := PushResultsToS3(ctx, cd, transport, s3Param, jobName, jobSHA, jobID)
+		if err != nil {
+			return nil, fmt.Errorf("PushResultToS3 did not finished: %w", err)
```
incorporate the job ID in the Errorf to distinguish in which job the error took place.

walterchris commented on plugins/postjobexecutionhooks/pushtoS3/pushtoS3.go:
```diff
@@ -11,32 +15,105 @@ import (
+		if err != nil {
+			return nil, err
+		}
+		// If awsFile was set to default
```
unintend to match code

walterchris commented on plugins/postjobexecutionhooks/pushtoS3/pushtoS3.go:
```diff
@@ -11,32 +15,105 @@ import (
+			return nil, err
+		}
+		// If awsFile was set to default
+	} else if s3Param.awsFile == "" {
```
shouldn't be an `else` be enough?

walterchris commented on plugins/postjobexecutionhooks/pushtoS3/pushtoS3.go:
```diff
@@ -11,32 +15,105 @@ import (
+		if !exists {
+			return fmt.Errorf("awsProfile does not exist!")
+		}
+	} else if awsProfile == "" {
```
just else

walterchris commented on plugins/postjobexecutionhooks/pushtoS3/pushtoS3.go:
```diff
@@ -11,32 +15,105 @@ import (
-func (n *PushtoS3) Run(ctx context.Context, parameters interface{}, cd client.ClientDescriptor, transport transport.Transport, rundata map[int]client.RunData) (interface{}, error) {
-	fmt.Println("Started the postjobplugin PushtoS3")
+	// Retrieving the parameter
+	var s3Param PushtoS3 = parameter.(PushtoS3)
```
In Camel case that would be PushToS3? In general.

walterchris replied:
Also that writting confuses me - is parameter.(PushtoS3) the same as parameter['PushtoS3'] ?

llogen replied:
I am doing a type assertion on the interface parameter to the PushToS3 struct, so i can pass is directly as the needed data type. parameter['PushtoS3'] does not work.

walterchris commented on plugins/postjobexecutionhooks/pushtoS3/s3.go:
```diff
@@ -22,243 +23,163 @@ import (
+		resultDesc := jobName + ". Test-Result:"
+		binaryDesc := jobName + ". Binary in S3 Bucket:"
+		// Parse the status resp for the binary url to update the binary status
+		regex := "https://" + parameter.s3Bucket + `[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)`
```
upload path?

llogen replied:
I don't know exactly what you mean.

walterchris replied:
Isn't the upload path fixed? So can't this be included?

llogen replied:
No the upload path in the contest-client is referring to the test results, because it uploads it. The path that is in the URL is the one that is referring to the binaries.

llogen committed:
plugins/postjobexecutionhooks/pushtoS3:
Reworked some minor things: Exported struct variables, corrected camel case, removed unnecessary else if's, incorporate the jobID in the error msg and added a TODO comment.

Signed-off-by: llogen <christophlange95@googlemail.com>

llogen committed:
pushtoS3:
Added Unit Testing and removed a bug in this plugin.

Signed-off-by: llogen <christophlange95@googlemail.com>

llogen committed:
plugins/postjobexecutionhooks/pushtoS3:
Added another unit test with two testdata files. Refactored the code of PushReportsToS3 and did some minor changes.

Signed-off-by: llogen <christophlange95@googlemail.com>

llogen committed:
Merge branch 'main' into fix/reworkPushtoS3
